File: experimental/haskell/card.hs
{-|
  This is the haskell-port of heckbert's card raytracer.
  Special thanks to Paul Heckbert

  See: http://tproger.ru/translations/business-card-raytracer

  12-May, 2018 : mayton - In beginning...
  28-May, 2018 : mayton - Fix for 'sum', Add console tests.
-}

import System.IO
import Text.Printf

width = 512 
    
height = 512
    
g = [0x0003C712,
     0x00044814,
     0x00044818,
     0x0003CF94,
     0x00004892,
     0x00004891,
     0x00038710,
     0x00000010,
     0x00000010]

data Vector = Vector { x :: Double, 
                       y :: Double, 
                       z :: Double } deriving (Show, Eq)


sprod :: Vector -> Vector -> Double
sprod v1 v2 = (x v1) * (x v2) + (y v1) * (y v2) + (z v1) * (z v2)

{- Vector sum of vector V1 and V2 -}
vsum :: Vector -> Vector -> Vector
vsum v1 v2 = (Vector 
               ((x v1) + (x v2)) 
               ((y v1) + (y v2)) 
               ((z v1) + (z v2)))

{--}
vprod :: Vector -> Vector -> Vector
vprod v1 v2 = (Vector
               (((y v1) * (z v2) - (z v1) * (y v2)))
               (((z v1) * (x v2) - (x v1) * (z v2)))
               (((x v1) * (y v2) - (y v1) * (x v2))))


prod :: Vector -> Double -> Vector
prod v r = (Vector (r * (x v)) (r * (y v)) (r * (z v)))

{- Normalyze vector V -}
norm :: Vector -> Vector
norm v = prod v ( 1.0 / (sqrt (sprod v v) ) )

v1 = (Vector 3.0 4.0 5.0)
v2 = (Vector 4.0 5.0 6.0)


{-squareloop x :: Integer = 
  do
    putStrLn ("x = "++ x)
    squareloop (x - 1) 
-}
main = do
  putStrLn ("v1 = " ++ (show v1))
  putStrLn ("v2 = " ++ (show v2))
  putStrLn ("norm(v1) = " ++ (show (norm v1)))
  putStrLn ("vprod(v1,v2) = " ++ (show (vprod v1 v2)))
  putStrLn ("sprod(v1,v2) = " ++ (show (sprod v1 v2)))
  putStrLn ("prod(v1,2.0) = " ++ (show (prod v1 2.0)))
  putStrLn ("vsum(v1,v2) = " ++ (show (vsum v1 v2)))

File: experimental/erlang/card.erl
% Special thanks to Paul Heckbert
%
% http://tproger.ru/translations/business-card-raytracer
%
% Erlang port
% 
% 07-May-2018 - (mayton) In beginning...
% 07-Sep-2019 - continue
% 12-Mar-2021 - Add recursive route
% 13-Mar-2021 - Add sampler, tracer


-module(card).

%-include_lib("stdlib/src/math.erl").
%-include_lib("eunit/include/eunit.hrl").
%-include_lib("stdlib/src/random.hrl").

-export([main/0]).

-define(WIDTH, 5).
-define(HEIGHT, 7).

-record(vector, {x :: double,
                 y :: double,
                 z :: double}).

g_data() ->
  [
            2#00111100011100010010,
            2#01000100100000010100,
            2#01000100100000011000,
            2#00111100111110010100,
            2#00000100100010010010,
            2#00000100100010010001,
            2#00111000011100010000,
            2#00000000000000010000,
            2#00000000000000010000
].


newvec(X,Y,Z) -> #vector{x=X,y=Y,z=Z}.

sum(A, B) -> newvec(A#vector.x + B#vector.x, A#vector.y + B#vector.y, A#vector.z + B#vector.z).

prod(A, K) -> newvec(A#vector.x * K, A#vector.y * K, A#vector.z * K).

sprod(A, B) -> A#vector.x * B#vector.x + A#vector.y * B#vector.y + A#vector.z * B#vector.z.

vprod(A, R) -> newvec(A#vector.y * R#vector.z - A#vector.z * R#vector.y,
                      A#vector.z * R#vector.x - A#vector.x * R#vector.z,
                      A#vector.x * R#vector.y - A#vector.y * R#vector.x).

norm(A) -> prod(A, 1.0 / math:sqrt(sprod(A,A))).

% TODO: Should be completed

tracer() -> 1.

%subPixelRoute(X, Y, R) ->
%  T = #vector {X = a * (random:uniform() - 0.5) * 99 + b * (Random() - .5) * 99}
  


sampler(Ovect, Dvect) -> 
  Mint = tracer(),
  ok.


rayJitter(Avec, Bvec, Rcount) when Rcount > 0 ->
  RandomVal = random:uniform(),
  % Vector t = 
  %       a * (random() - 0.5) * 99.0 + 
  %       b * (random() - 0.5) * 99.0
  Tvec = sum(99.0 * prod(Avec, random:uniform() - 0.5),
             99.0 * prod(Bvec, random:uniform() - 0.5)),
  % p = sampler(Vector(17, 16, 8) + t, 
  %             !(t * -1 + (a * (Random() + x) + b * (y + Random()) + c) * 16)) * 3.5 + p;
  Ovec = sum(newvec(17.0,16.0,8.0),Tvec)
  Dvec = newvec(0.0,0.0,0.0),
  Pvec = sampler(Ovec, Dvec),
  rayJitter(Avec, Bvec, Rcount - 1).


pixelRoute(X, Y) when X < ?WIDTH , Y < ?HEIGHT -> 
  Gvec = norm(newvec(-6.0,-16.0,0.0)),
  Avec = prod(norm(newvec(0.0,0.0,1.0)),0.002),
  Bvec = norm(prod(vprod(Gvec,Avec),0.002)),
  Cvec = sum((prod(sum(Avec,Bvec),-256.0)),Gvec),
  %io:fwrite("(~p,~p,~p);~n",[X,Y,F]),
  rayJitter(Avec,Bvec, 64),
  pixelRoute(X + 1, Y);

pixelRoute(X, Y) when X == ?WIDTH , Y < ?HEIGHT -> 
  pixelRoute(0, Y + 1);

pixelRoute(X, Y) -> ok.

main() -> 
   io:fwrite("P6 ~p ~p 255 ",[?WIDTH,?HEIGHT]),
   pixelRoute(0,0).
